src/awkward/highlevel.py

```diff
@@ -1062,13 +1094,30 @@ def __getitem__(self, where):
         have the same dimension as the array being indexed.
         """
         with ak._errors.SlicingErrorContext(self, where):
-            return wrap_layout(
-                prepare_layout(self._layout[where]),
+            # Handle named axis
+            (_, ndim) = self._layout.minmax_depth
+            named_axis = _get_named_axis(self)
+            where = _normalize_named_slice(named_axis, where, ndim)
+
+            NamedAxis.mapping = named_axis
+
+            out = wrap_layout(
+                prepare_layout(self._layout._getitem(where, NamedAxis)),
                 self._behavior,
                 allow_other=True,
                 attrs=self._attrs,
             )
 
+            if NamedAxis.mapping:
+                return ak.operations.ak_with_named_axis._impl(
+                    out,
+                    named_axis=NamedAxis.mapping,
+                    highlevel=True,
+                    behavior=None,
+                    attrs=None,
+                )
```

Comment:
I haven't thought about this deeply, but I think it would be better to defer the first `wrap_layout` until the `else` branch of `if NamedAxis.mapping`. Then, `with_named_axis` doesn't need to unwrap the high-level array (and, thus, you'll need to pass in the `attrs` and `behavior` arguments to `with_named_axis`.

Reply:
I'll try this out - less unwrapping + wrapping makes sense. This was a little fiddly part in the code where I wanted to make sure that the return type stay the same compared to awkward-array/main. But I'm sure your suggestion would not alter the return type, so great!